plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -735,3 +735,188 @@ func haveSame(a1, a2 []string) bool {
+	node, err := n.info.GetNodeInfo(nodeName)
+	if err != nil {
+		return false, err
+	}
+	allPods, err := n.podLister.List(labels.Everything())
+	if err != nil {
+		return false, err
+	}
+	return nodeMatchPodAffinityAntiAffinity(pod, nodeInfo.Pods(), node, allPods), nil
+}
+
+// checkAnyPodThatMatchPodAffinityTerm checks if any of existing pods on the node can match the specific podAffinityTerm.
+func checkAnyPodThatMatchPodAffinityTerm(pod *api.Pod, existingPods []*api.Pod, node *api.Node, podAffinityTerm api.PodAffinityTerm) (bool, error) {
+	labelSelector, err := unversioned.LabelSelectorAsSelector(podAffinityTerm.LabelSelector)
+	if err != nil {
+		return false, err
+	}
+
+	names := priorityutil.GetNamespacesFromPodAffinityTerm(pod, podAffinityTerm)
+	filteredPods := priorityutil.FilterPodsByNameSpaces(names, existingPods)
+	// return true if any pod matches the podAffinityTerm
+	for _, filteredPod := range filteredPods {
+		if labelSelector.Matches(labels.Set(filteredPod.Labels)) {
+			return priorityutil.IfNodeHasTopologyKey(node, podAffinityTerm.TopologyKey), nil
+		}
+	}
+	return false, nil
+}
+
+// Checks whether the given node has pods which satisfy all the required pod affinity scheduling rules.
+// If node has pods which satisfy all the required pod affinity scheduling rules then return true.
+func checkNodeMatchesHardPodAffinity(pod *api.Pod, existingPods []*api.Pod, node *api.Node, allPods []*api.Pod, podAffinity *api.PodAffinity) bool {
+	var podAffinityTerms []api.PodAffinityTerm
+	if len(podAffinity.RequiredDuringSchedulingIgnoredDuringExecution) != 0 {
+		podAffinityTerms = podAffinity.RequiredDuringSchedulingIgnoredDuringExecution
+	}
+	// TODO: Uncomment this block when implement RequiredDuringSchedulingRequiredDuringExecution.
+	//if len(podAffinity.RequiredDuringSchedulingRequiredDuringExecution) != 0 {
+	//	podAffinityTerms = append(podAffinityTerms, podAffinity.RequiredDuringSchedulingRequiredDuringExecution...)
+	//}
+
+	for _, podAffinityTerm := range podAffinityTerms {
+		podAffinityTermMatches, err := checkAnyPodThatMatchPodAffinityTerm(pod, existingPods, node, podAffinityTerm)
+		// if err or podAffinityTerm match failed then return false otherwise continue
+		if err != nil {
+			glog.V(10).Infof("Cannot schedule pod %+v onto node %v, an error ocurred when checking existing pods on the node for PodAffinityTerm %v err: %v",
+				podName(pod), node.Name, podAffinityTerm, err)
+			return false
+		}
+
+		if !podAffinityTermMatches {
+			// Check if it is in special case that the requiredDuringScheduling affinity requirement can be disregarded.
+			// If the requiredDuringScheduling affinity requirement matches a pod's own labels, and there are no other such pods
+			// anywhere, then disregard the requirement.
+			labelSelector, err := unversioned.LabelSelectorAsSelector(podAffinityTerm.LabelSelector)
+			if err != nil || !labelSelector.Matches(labels.Set(pod.Labels)) {
```

Comment:
Probably this should also check that `Namespaces == nil || Namespaces contains pod.Namespace`
(in other words, rule should be "If the requiredDuringScheduling affinity requirement matches a pod's own labels and namespace, and there are no other such pods anywhere, then disregard the requirement.")